# 4.5 - Data Consistency Check - Part 2

### This script contains the following points:

* 01 - Importing Libraries
* 02 - Importing Data
* 03 - Mixed Data Types
* 04 - Missing Values
* 05 - Duplicates Values
* 06 - Tasks 4.5 - Data Consistency Check
     * a) df_prods
     * b) df_ords
* 07 - Exporting Data

---

## 01. Importing Libraries

In [1]:
# Import libraries

import numpy as np
import pandas as pd
import os

---

## 02. Importing Data

In [2]:
path = r'/Users/juanigalvalisi/01-07-2022 - Instacart Basket Analysis'

In [3]:
# Import CSVs

df_prods = pd.read_csv(r'/Users/juanigalvalisi/01-07-2022 - Instacart Basket Analysis/02 - Data/Original Data/products.csv', index_col = False)

In [4]:
df_ords = pd.read_csv(r'/Users/juanigalvalisi/01-07-2022 - Instacart Basket Analysis/02 - Data/Prepared Data/orders_wrangled.csv', index_col = False)

In [5]:
# Check the output

df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_last_order  
0                  8              NaN  
1                  7             15.0  
2                 12             21.0  
3                  7             29.0  
4                 15             28.0

In [6]:
# Check the output II

df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

---

## 03 - Mixed Data Types

In [7]:
# Create a dataframe

df_test = pd.DataFrame()

In [8]:
# Create a mixed type column

df_test['mix'] = ['a', 'b', 1, True]

In [9]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [10]:
# Look for mixed types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [11]:
# Fix mixed types (you can choose any data type you want)

df_test['mix'] = df_test['mix'].astype('str')

---

## 04 - Missing Values

In [12]:
# Find Missing Values

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
# Script to view these 16 missing values. Create a new subset containing only these values in question

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [15]:
# How to solve these NaN values? (ONLY FOR NUMERICAL VALUES)

# IMPUTING: df.describe() to find the mean

df_nan.describe()

product_id    aisle_id  department_id     prices
count     16.000000   16.000000      16.000000  16.000000
mean    6684.000000   89.937500      10.937500  13.012500
std    12836.665242   33.731229       4.639953   3.881731
min       34.000000   26.000000       1.000000   1.200000
25%      459.250000   70.750000       7.750000  12.175000
50%     2413.000000   98.500000      11.500000  13.650000
75%     3872.750000  120.000000      14.500000  14.425000
max    40440.000000  126.000000      16.000000  20.900000

In [16]:
# To replace these NaN values with the mean

#df['column_with_missings'].fillna(mean value, inplace = True)

In [17]:
# To replace these NaN values with the median

#df['column_with_missings'].fillna(median value, inplace = True)

In [18]:
# Remove the missing values entirely or filter out the ones that aren’t missing
# into a subset dataframe and continue your analysis with this new dataframe

# 1) You can compare the number of rows in your current dataframe with the number
# in your subset once the missing rows have been removed:

df_prods.shape

(49693, 5)

In [19]:
# 2) Create a new dataframe

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [20]:
# 3) Check if the rows are decreased

df_prods_clean.shape

(49677, 5)

In [21]:
# OVERWRITING instead of creating a new dataframe

# Remove the NaN values (ALL)

df_prods.dropna(inplace = True)

# Remove the NaN values (PARTICULAR COLUMN)

df_prods.dropna(subset = ['product_name'], inplace = True)

---

## 05 - Duplicates

In [22]:
# Looking for duplicates.

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [23]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [24]:
# Addressing duplicates
# 1) Check the number of rows

df_prods_clean.shape

(49677, 5)

In [25]:
# 2) Create a new dataframe that doesn’t include the duplicates you just identified and delete them

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [26]:
# 3) Check the elimination of duplicates

df_prods_clean_no_dups.shape

(49672, 5)

---

## 06 - Tasks 4.5 - Data Consistency Check

#### a) df_prods

In [27]:
# 2) Run the df.describe() function on your df_prods dataframe. Share in a markdown cell
# whether anything about the data looks off or should be investigated further.

In [28]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

#### Generally speaking, the data seems to be fine except for the maximum price record, which could be an outlier.

# Change PRICES 14999 Y 99999

In [29]:
#df_prods_clean_no_dups

#### b) df_ords

In [30]:
# 3) Check for mixed-type data in your df_ords dataframe. If you find mixed-type data, fix it.
# The column in question should contain observations of a single data type.

In [31]:
# Looking mixed-type data in all our dataframe.

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

#### There are no mixed-type data within the df_ords dataframe.

In [32]:
# 4) Run a check for missing values in your df_ords dataframe. In a markdown cell,
# report your findings and propose an explanation for any missing values you find.

In [33]:
# Looking for missing values in all our dataframe.

df_ords.isnull().sum()

Unnamed: 0                 0
order_id                   0
user_id                    0
order_number               0
orders_day_of_week         0
order_hour_of_day          0
days_last_order       206209
dtype: int64

In [34]:
# Filtering by missing values withing the 'days_last_order' variable.

df_ords_nan = df_ords[df_ords['days_last_order'].isnull() == True]

In [35]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_last_order  
0                        8              NaN  
11                      11              NaN  
26                      14              NaN  
39                      11              NaN  
45                      12              NaN  
...                    ...              ...  
3420930                 12              NaN  
3420934                 18              NaN  
3421002                 19              NaN  
3421019                 15              NaN  
3421069                 11              NaN  

[206209 rows x 7 columns]

#### There are 206.209 missing values only within the 'days_last_order' variable. These missing values may exist because it was the first time the customers made a purchase.

In [36]:
# 5) Address the missing values using an appropriate method.
# In a markdown cell explain why you used your method of choice.

In [37]:
# Defining our new dataframe.

df_ords_clean = df_ords

In [38]:
# Creating a new column where True represents a new customer and False means an old customer.

df_ords_clean['new_customer'] = df_ords['days_last_order'].isnull() == True

In [39]:
df_ords_clean

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_last_order  new_customer  
0                        8              NaN          True  
1                        7             15.0         False  
2                       12             21.0         False  
3                        7             29.0         False  
4                       15             28.0         False  
...                    ...              ...           ...  
3421078                 18             29.0         False  
3421079                 10             30.0         False  
3421080                 12             18.0         False  
3421081                 12              7.0         False  
3421082                 14             30.0         False  

[3421083 rows x 8 columns]

#### We could handle this situation by creating a new column under the boolean data type with the aim of indicating whether they are new (True) customers or not (False). This new variable will act as a supplement to the days_last_order column.


In [40]:
# 6) Run a check for duplicate values in your df_ords data. In a markdown cell,
# report your findings and propose an explanation for any duplicate values you find

In [41]:
df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [42]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_last_order, new_customer]
Index: []

#### There are no duplicates values within our df_ords_dupl dataframe.

In [43]:
# Change orders_day_of_week values for days of the week



---

## 07 - Exporting Data

In [44]:
# Export your final, cleaned df_prods and df_ords data as '.csv' files in
# your 'Prepared Data' folder and give them appropriate, succinct names

In [45]:
# Exporting df_prods as 'products_checked.csv'

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 - Data', 'Prepared Data', 'products_checked.csv'))

In [46]:
# Exporting df_ords as 'orders_checked.csv'

df_ords_clean.to_csv(os.path.join(path, '02 - Data', 'Prepared Data', 'orders_checked.csv'))